firstpartydata/first_party_data.go

```diff
@@ -159,13 +160,49 @@ func ResolveFPD(bidRequest *openrtb2.BidRequest, fpdBidderConfigData map[openrtb
 		}
 		resolvedFpdConfig.Site = newSite
 
+		newDevice, err := resolveDevice(fpdConfig, bidRequest.Device)
+		if err != nil {
+			errL = append(errL, err)
+		}
+		resolvedFpdConfig.Device = newDevice
+
 		if len(errL) == 0 {
 			resolvedFpd[openrtb_ext.BidderName(bidderName)] = resolvedFpdConfig
 		}
 	}
 	return resolvedFpd, errL
 }
 
+// resolveDevice merges the device information from the FPD (First Party Data) configuration
+// with the device information provided in the bid request. It returns a new Device object
+// that contains the merged data.
+func resolveDevice(fpdConfig *openrtb_ext.ORTB2, bidRequestDevice *openrtb2.Device) (*openrtb2.Device, error) {
+	var fpdConfigDevice json.RawMessage
+
+	if fpdConfig != nil && fpdConfig.Device != nil {
+		fpdConfigDevice = fpdConfig.Device
+	}
+
+	if bidRequestDevice == nil && fpdConfigDevice == nil {
+		return nil, nil
+	}
+
+	var newDevice *openrtb2.Device
+	if bidRequestDevice != nil {
+		newDevice = ptrutil.Clone(bidRequestDevice)
+	} else {
+		newDevice = &openrtb2.Device{}
+	}
+
+	if fpdConfigDevice != nil {
+		if err := jsonutil.MergeClone(newDevice, fpdConfigDevice); err != nil {
+			return nil, formatMergeCloneError(err)
+		}
+	}
+
+	return newDevice, nil
+}
```

Comment:
Can you please add device validation before returning it? 
Logic should be the same as in [func validateDevice(device *openrtb2.Device)](https://github.com/prebid/prebid-server/blob/17f40d133486d8b65074cc4ad0b3dc3ab0d8e567/endpoints/openrtb2/auction.go#L1328). 
This function is not available here, so copying it is ok. 
Also please add a unit test for this.

Reply:
Added in 7abb2a88ac1cd09c59f047b3bf7735f321488868